File: 2023/Runner.jl
# Copyright 2023 Google LLC
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

# Advent of Code runner for Julia.

module Runner
using Printf

function run_part(func, input, expected)
  stats = @timed func(input)
  Result(stats.value, expected, stats.time, stats.bytes)
end

function run_module(mod, filenames; verbose=false)
  name = nameof(mod)
  ok = true
  for fname in filenames
    lines = readlines(fname == "-" ? stdin : fname)
    len = length(lines)
    expect = expectedfor(fname)
    for func in [mod.part1, mod.part2]
      expected = get(expect, string(func), "")
      if verbose
        println(stderr, "Running $name $func on $fname ($len lines)")
      end
      res = run_part(func, lines, expected)
      println("$func: $(res.got)")
      if verbose
        time = format_seconds(res.time_sec)
        bytes = Base.format_bytes(res.allocated_bytes)
        print_message(stderr, res)
        println(stderr, "$func took $time and $bytes on $fname")
        println(stderr, "="^40)
      end
    end
  end
  return ok
end

expectedfile(inputfile) = replace(inputfile, r"(.*)\.txt$" => s"\1.expected")

function expectedfor(inputfile)
  expectfile = expectedfile(inputfile)
  if expectfile != inputfile && isfile(expectfile)
    open(expectfile) do ef
      map(m -> m.captures[1] => replace(m.captures[2], r"\\n" => "\n"),
        map(l -> match(r"(part\d):\s*(.*)", l), readlines(ef)) |>
        filter(!isnothing))
    end |> Dict{String, String}
  else
    Dict{String, String}()
  end
end

function format_seconds(sec)
  if sec >= 60 * 60
    (m, s) = divrem(sec, 60)
    (h, m) = divrem(m, 60)
    @sprintf("%d:%02d:%02f", h, m, s)
  elseif sec >= 60
    (m, s) = divrem(sec, 60)
    @sprintf("%d:%02.3f", m, s)
  elseif sec >= 1
    @sprintf("%.3fs", sec)
  elseif sec >= 0.001
    @sprintf("%.3fms", sec*1_000)
  else
    @sprintf("%.3fµs", sec*1_000_000)
  end
end

struct Result
  got::Any
  want::String
  time_sec::Float64
  allocated_bytes::Int
end

const OUTCOME_SYMBOLS = Dict(:success => "✅",
  :failure => "❌",
  :unknown => "❓",
  :todo => "❗")
const OUTCOME_BG_COLOR = Dict(:success => :light_green,
  :failure => :light_red,
  :unknown => :magenta,
  :todo => :cyan)
function print_message(io, res::Result)
  o = outcome(res)
  sign = OUTCOME_SYMBOLS[o]
  msg = if o == :success
    "got $(res.got)"
  elseif o == :failure
    "got $(res.got), want $(res.want)"
  elseif o == :unknown
    "got $(res.got)"
  elseif o == :todo && res.want == ""
    "implement it"
  elseif o == :todo
    "implement it, want $(res.want)"
  end
  bg = OUTCOME_BG_COLOR[o]
  buf = IOBuffer()
  printstyled(IOContext(buf, :color => true),
    uppercase(string(o)),
    color=bg,
    reverse=true)
  colored = String(take!(buf))
  println(io, "$sign $colored $msg")
end

function outcome(res::Result)
  if res.got == :TODO
    :todo
  elseif isempty(res.want)
    :unknown
  elseif string(res.got) == res.want
    :success
  else
    :failure
  end
end

end

macro run_if_main()
  srcfile = QuoteNode(__source__.file)
  mod = __module__
  :(if abspath(PROGRAM_FILE) == string($srcfile)
    args = ARGS
    verbose = false
    if isempty(ARGS)
      args = ["-"]
    elseif first(ARGS) in ["-v", "--verbose"]
      verbose = true
      args = args[2:end]
    end
    ok = Runner.run_module($mod, args, verbose=verbose)
    exit(ok ? 0 : 1)
  end)
end

File: lang/go/runner.go
// Copyright 2023 Google LLC
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

// runner.go provides runMain and other support functions to read input files,
// run an Advent of Code solution, and log the results.
// Symlink to this file from the directory with the solution and run with
// % go run dayX/*.go -v path/to/input.*.txt
// which enables running everything in a single main package without modules.

package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
	"time"
)

type Part func(lines []string) string

var (
	verbose = false
)

func runMain(part1, part2 Part) {
	log.SetFlags(log.Ltime)
	flag.BoolVar(&verbose, "verbose", false, "log time and status")
	flag.BoolVar(&verbose, "v", false, "log time and status")
	flag.Parse()
	files := flag.Args()
	if len(files) == 0 {
		files = []string{"-"} // read stdin
	}
	success := true
	for _, fname := range files {
		success = runFile(fname, part1, part2) && success
	}
	if success {
		os.Exit(0)
	}
	os.Exit(1)
}

func runFile(fname string, part1, part2 Part) bool {
	lines, err := readLines(fname)
	if err != nil {
		log.Fatal(err)
	}
	expect := readExpected(fname)
	p1 := execution{part: part1, partName: "part1", fileName: fname, lines: lines, expected: expect[0]}
	p2 := execution{part: part2, partName: "part2", fileName: fname, lines: lines, expected: expect[1]}
	success := p1.run()
	success = p2.run() && success
	return success
}

func readLines(fname string) ([]string, error) {
	var f io.Reader
	if fname == "-" {
		f = os.Stdin
	} else {
		file, err := os.Open(fname)
		if err != nil {
			return nil, fmt.Errorf("error reading %s: %v", fname, err)
		}
		defer file.Close()
		f = file
	}
	lines := make([]string, 0)
	s := bufio.NewScanner(f)
	for s.Scan() {
		lines = append(lines, s.Text())
	}
	if err := s.Err(); err != nil {
		return nil, fmt.Errorf("error reading lines from %s: %v", fname, err)
	}
	return lines, nil
}

func readExpected(inputfname string) [2]string {
	res := [2]string{"", ""}
	if !strings.HasSuffix(inputfname, ".txt") {
		return res
	}
	efname := strings.TrimSuffix(inputfname, "txt") + "expected"
	f, err := os.Open(efname)
	if err != nil {
		return res
	}
	defer f.Close()
	s := bufio.NewScanner(f)
	for s.Scan() {
		line := s.Text()
		if strings.HasPrefix(line, "part1: ") {
			res[0] = strings.TrimPrefix(line, "part1: ")
		}
		if strings.HasPrefix(line, "part2: ") {
			res[1] = strings.TrimPrefix(line, "part2: ")
		}
	}
	for i, t := range res {
		res[i] = strings.ReplaceAll(t, "\\n", "\n")
	}
	return res
}

type execution struct {
	part     Part
	partName string
	fileName string
	lines    []string
	expected string
}

func (e execution) run() bool {
	if verbose {
		log.Printf("Running %s %s on %s (%d lines)", dayName, e.partName, e.fileName, len(e.lines))
	}
	start := time.Now()
	l := make([]string, len(e.lines))
	copy(l, e.lines)
	res := e.part(l)
	elapsed := time.Since(start)
	fmt.Printf("%s: %s\n", e.partName, res)
	if verbose {
		var msg string
		if res == e.expected {
			msg = fmt.Sprintf("✅ %s got %s", colored(colorSuccess, "SUCCESS"), res)
		} else if res == "TODO" {
			msg = fmt.Sprintf("❗ %s implement it", colored(colorTodo, "TODO"))
			if e.expected != "" {
				msg += fmt.Sprintf(", want %s", e.expected)
			}
		} else if e.expected == "" {
			msg = fmt.Sprintf("❓ %s got %s", colored(colorUnknown, "UNKNOWN"), res)
		} else {
			msg = fmt.Sprintf("❌ %s got %s, want %s", colored(colorFailure, "FAILURE"), res, e.expected)
		}
		log.Println(msg)
		log.Printf("%s took %s on %s", e.partName, elapsed, e.fileName)
	}
	return res == e.expected
}

const (
	colorSuccess = "30;102" // black on bright green
	colorFailure = "30;101" // black on bright red
	colorUnknown = "30;103" // black on bright yellow
	colorTodo    = "30;106" // black on bright cyan
)

func colored(color, s string) string {
	return fmt.Sprintf("\x1B[%sm%s\x1B[0m", color, s)
}
